src/XrdHttp/XrdHttpProtocol.cc

```diff
@@ -283,124 +285,100 @@ XrdProtocol *XrdHttpProtocol::Match(XrdLink *lp) {
 
 int XrdHttpProtocol::GetVOMSData(XrdLink *lp) {
   TRACEI(DEBUG, " Extracting auth info.");
+  char bufname[256];
+  int mape;
+  XrdCryptoX509Chain chain;
 
-  X509 *peer_cert;
+  if (!myCryptoFactory->X509ParseStack()((void*) ssl, &chain)) {
+    TRACEI(DEBUG, "No certificates found in peer chain.");
+    return 0;
+  }
 
-  // No external plugin, hence we fill our XrdSec with what we can do here
-  peer_cert = SSL_get_peer_certificate(ssl);
-  TRACEI(DEBUG, " SSL_get_peer_certificate returned :" << peer_cert);
-  ERR_print_errors(sslbio_err);
+  const char * dn = chain.EECname();
 
-  if (peer_cert) {
-    char bufname[256];
-    
-    // Add the original DN to the moninfo. Not sure if it makes sense to parametrize this or not.
-    if (SecEntity.moninfo) free(SecEntity.moninfo);
-    
-    // The original DN is not so univoque. In the case of a proxy this can be either
-    // the issuer name or the subject name
-    // Here we try to use the one that is known to the user mapping
-    
-    
-    
-    if (servGMap) {
-      int mape;
-      
-      SecEntity.moninfo = X509_NAME_oneline(X509_get_issuer_name(peer_cert), NULL, 0);
-      TRACEI(DEBUG, " Issuer name is : '" << SecEntity.moninfo << "'");
-      
-      mape = servGMap->dn2user(SecEntity.moninfo, bufname, sizeof(bufname), 0);
-      if ( !mape && SecEntity.moninfo[0] ) {
-        TRACEI(DEBUG, " Mapping name: '" << SecEntity.moninfo << "' --> " << bufname);
-        if (SecEntity.name) free(SecEntity.name);
-        SecEntity.name = strdup(bufname);
-      }
-      else {
-        TRACEI(ALL, " Mapping name: '" << SecEntity.moninfo << "' Failed. err: " << mape);
-        
-        // Mapping the issuer name failed, let's try with the subject name
-        if (SecEntity.moninfo) free(SecEntity.moninfo);
-        SecEntity.moninfo = X509_NAME_oneline(X509_get_subject_name(peer_cert), NULL, 0);
-        TRACEI(DEBUG, " Subject name is : '" << SecEntity.moninfo << "'");
-        
-        mape = servGMap->dn2user(SecEntity.moninfo, bufname, sizeof(bufname), 0);
-        if ( !mape ) {
-          TRACEI(DEBUG, " Mapping name: " << SecEntity.moninfo << " --> " << bufname);
-          if (SecEntity.name) free(SecEntity.name);
-          SecEntity.name = strdup(bufname);
-        }
-        else {
-          TRACEI(ALL, " Mapping name: " << SecEntity.moninfo << " Failed. err: " << mape);
-        }
-      }
-      
+  if (!dn) {
```

Comment:
Is falling back to OpenSSL safe/reasonable?  That is, is there a setup where a peer certificate is available and works but the `XrdCrypto` library doesn't?

I mention this because this code will definitely do the wrong thing for proxies.

Reply:
If we don't do this then simple certificates don't work. The EEC for a plain certificate is the issuer and what you want is the subject in this case.

Reply:
No... the EEC for a plain certificate is still the correct subject (at least, this was done correctly for my testing).

The algorithm for discovering the EEC is (basically) to order the chain and, starting at the end, walking up the chain.  The EEC is the last certificate discovered that is _not_ marked as a CA (a well-defined attribute within X509).

Did your testing reveal issues here?

Reply:
Yes, for my plain simple certificate issued by the CERN GridCA the EEC was the CERN GridCA. There was only one certificate in the chain and this was the CA one. Also looking at the description of SSL_get_peer_cert_chain there is an explicit mentioning that one should use SSL_get_peer_certificate:
https://www.openssl.org/docs/man1.0.2/man3/SSL_get_peer_cert_chain.html

Reply:
Interesting!  I wonder what my test worked then.

Let me investigate.  I'd like to unify this to a single library function instead of splitting over two approaches.